src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -0,0 +1,244 @@
+package is.hail.distributedmatrix
+
+import is.hail.utils._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed._
+
+import scala.reflect.ClassTag
+
+object BlockMatrixIsDistributedMatrix extends DistributedMatrix[BlockMatrix] {
+  type M = BlockMatrix
+
+  def cache(m: M): M = m.cache()
+
+  def from(irm: IndexedRowMatrix, dense: Boolean = true): M =
+    if (dense) irm.toBlockMatrixDense()
+    else irm.toBlockMatrix()
+  def from(bm: BlockMatrix): M = bm
+  def from(cm: CoordinateMatrix): M =
+    cm.toBlockMatrix()
+
+  def transpose(m: M): M = m.transpose
+  def diagonal(m: M): Array[Double] = {
+    val rowsPerBlock = m.rowsPerBlock
+    val colsPerBlock = m.colsPerBlock
+
+    // check if the intervals [i, i+rows) and [j, j+rows) overlap,
+    // meaning this block contains the diagonal.
+    def containsDiagonal(coord: (Int, Int)): Boolean = coord match {
+      case (blocki, blockj) =>
+        val i = blocki * rowsPerBlock
+        val j = blockj * colsPerBlock
+        val i2 = i + rowsPerBlock
+        val j2 = i + colsPerBlock
+
+        i <= j && j < i2 ||
+        i < j2 && j2 < i2 ||
+        j <= i && i < j2 ||
+        j < i2 && i2 < j2
+    }
+
+    def diagonal(block: Matrix): Array[Double] = {
+      val length = math.min(block.numRows,block.numCols)
+      val diagonal = new Array[Double](length)
+      var i = 0
+      while (i < length) {
+        diagonal(i) = block(i, i)
```

Comment:
The matrix diagonal won't necessarily lie on the diagonal of the block.  Consider rowsPerBlock = 10, colsPerBlock = 100, and consider block (1, 0).  The diagonal starts at (0, 10).

It will if rowsPerBlock = colsPerBlock.  Do have a case where we want to support non-square blocks?